pkg/minikube/registry/drvs/hyperv/vswitch.go

```diff
@@ -0,0 +1,212 @@
+	if physical {
+		cmdlet = append(cmdlet, "-Physical")
+	}
+	cmd := []string{strings.Join(cmdlet, " ")}
+	if condition != "" {
+		cmd = append(cmd, fmt.Sprintf("Where-Object {%s}", condition))
+	}
+	cmd = append(cmd, "Select-Object -Property InterfaceGuid, InterfaceDescription")
+	stdout, err := cmdOut(fmt.Sprintf("ConvertTo-Json @(%s)", strings.Join(cmd, " | ")))
+	if err != nil {
+		return nil, err
+	}
+
+	var adapters []netAdapter
+	err = json.Unmarshal([]byte(strings.TrimSpace(stdout)), &adapters)
+	if err != nil {
+		return nil, err
+	}
+
+	return adapters, nil
+}
+
+func getVMSwitch(condition string) ([]vmSwitch, error) {
+	cmd := []string{"Hyper-V\\Get-VMSwitch"}
+	if condition != "" {
+		cmd = append(cmd, fmt.Sprintf("Where-Object {%s}", condition))
+	}
+	cmd = append(cmd, "Select-Object -Property Name, NetAdapterInterfaceGuid")
+	stdout, err := cmdOut(fmt.Sprintf("ConvertTo-Json @(%s)", strings.Join(cmd, " | ")))
+	if err != nil {
+		return nil, err
+	}
+
+	var vmSwitches []vmSwitch
+	err = json.Unmarshal([]byte(strings.TrimSpace(stdout)), &vmSwitches)
+	if err != nil {
+		return nil, err
+	}
+
+	return vmSwitches, nil
+}
+
+func findConnectedVMSwitch(adapterGUID string) (string, error) {
+	foundSwitches, err := getVMSwitch(fmt.Sprintf("($_.SwitchType -eq 2) -And ($_.NetAdapterInterfaceGuid -contains \"%s\")", adapterGUID))
+	if err != nil {
+		return "", err
+	}
+
+	if len(foundSwitches) > 0 {
+		return foundSwitches[0].Name, nil
+	}
+
+	return "", nil
+}
+
+func getOrderedAdapters() ([]netAdapter, error) {
+	// look for "Up" adapters and prefer physical LAN over other options
+	lanAdapters, err := getNetAdapters(true, "($_.Status -eq \"Up\") -And ($_.PhysicalMediaType -like \"*802.3*\")")
+	if err != nil {
+		return nil, err
+	}
+
+	upAdapters, err := getNetAdapters(true, "$_.Status -eq \"Up\"")
+	if err != nil {
+		return nil, err
+	}
+
+	var orderedAdapters []netAdapter
+	adapterGuids := map[string]interface{}{}
+
+	// all adapters will be checked in the following order:
+	//  1. Connected physical LAN adapters
+	//  2. Any other connected adapters
+	for _, adapter := range lanAdapters {
+		if _, ok := adapterGuids[adapter.InterfaceGUID]; !ok {
+			adapterGuids[adapter.InterfaceGUID] = nil
+			orderedAdapters = append(orderedAdapters, adapter)
+		}
+	}
+
+	for _, adapter := range upAdapters {
+		if _, ok := adapterGuids[adapter.InterfaceGUID]; !ok {
+			adapterGuids[adapter.InterfaceGUID] = nil
+			orderedAdapters = append(orderedAdapters, adapter)
+		}
+	}
+
+	return orderedAdapters, nil
+}
+
+func createVMSwitch(switchName string, adapter netAdapter) error {
+	err := cmd(fmt.Sprintf("Hyper-V\\New-VMSwitch -Name \"%s\" -NetAdapterInterfaceDescription \"%s\"", switchName, adapter.InterfaceDescription))
+	if err != nil {
+		return errors.Wrapf(err, "failed to create VM switch %s with adapter %s", switchName, adapter.InterfaceGUID)
+	}
+
+	return nil
+}
+
+func chooseSwitch(adapterName string) (string, error) {
```

Comment:
I wasn't expecting a method called `chooseSwitch` to create a switch. Could we refactor it so that it passes the necessary information back in case a switch needs to be created?

Maybe like:

```
switch, err := selectSwitch(adapter)
if err != nil {
  switch, err := createSwitch(adapter)
```

Reply:
Done